### ISSUE TITLE
[Breaking change]: C# overload resolution prefers span overloads which might not work in Expression lambdas

### ISSUE THREAD

**Issue by jjonescz** [dotnet/docs#43952]
### Description

C# 14 introduces new [built-in span conversions and type inference rules](https://github.com/dotnet/csharplang/issues/7905) making overloads with span parameters applicable in more scenarios.

However, Expression lambdas cannot be interpreted when they involve ref structs like Span and ReadOnlySpan. [For example](https://github.com/dotnet/runtime/issues/109757):

```cs
using System;
using System.Linq;
using System.Linq.Expressions;

Expression<Func<int[], int, bool>> exp = (array, num) => array.Contains(num);
exp.Compile(preferInterpretation: true); // fails at runtime in C# 14
```

The `array.Contains` binds to `Enumerable.Contains` in C# 13 but binds to `MemoryExtensions.Contains` in C# 14. The latter involves ReadOnlySpan and hence crashes when `Expression.Compile(preferInterpretation: true)` is called.

### Version

.NET 10 Preview 1

### Previous behavior

In C# 13 and earlier, an extension method taking a `ReadOnlySpan<T>` or `Span<T>` receiver is not applicable to a value of type `T[]`. Therefore, only non-span extension methods like the ones from the `System.Linq.Enumerable` class are usually bound inside Expression lambdas.

### New behavior

In C# 14 and later, methods with `ReadOnlySpan<T>` or `Span<T>` parameters can participate in type inference or be used as extension methods in more scenarios. This makes span-based methods like the ones from the `System.MemoryExtensions` class bind in more scenarios, including inside Expression lambdas where they will cause runtime exceptions when compiled with interpretation.

### Type of breaking change

- [ ] **Binary incompatible**: Existing binaries might encounter a breaking change in behavior, such as failure to load or execute, and if so, require recompilation.
- [ ] **Source incompatible**: When recompiled using the new SDK or component or to target the new runtime, existing source code might require source changes to compile successfully.
- [x] **Behavioral change**: Existing binaries might behave differently at run time.

### Reason for change

The C# language feature allows simplified API design and usage (e.g., one ReadOnlySpan extension method can apply to both spans and arrays).

### Recommended action

If you need to continue using Expression interpretation, you should make sure the non-span overloads are bound, e.g., by casting arguments to the exact types the method signature takes or calling the extension methods as explicit static invocations:

```cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;

M((array, num) => array.Contains(num)); // fails, binds to MemoryExtensions.Contains
M((array, num) => ((IEnumerable<int>)array).Contains(num)); // ok, binds to Enumerable.Contains
M((array, num) => array.AsEnumerable().Contains(num)); // ok, binds to Enumerable.Contains
M((array, num) => Enumerable.Contains(array, num)); // ok, binds to Enumerable.Contains

void M(Expression<Func<int[], int, bool>> e) => e.Compile(preferInterpretation: true);
```

### Feature area

Other (please put exact area in description textbox)

### Affected APIs

```cs
System.Linq.Expressions.Expression.Compile
```

---
[Associated WorkItem - 360646](https://dev.azure.com/msft-skilling/Content/_workitems/edit/360646)

**Comment by BillWagner** [comment]
Adding myself as an assignee here.

I need to publish https://github.com/dotnet/roslyn/blob/main/docs/compilers/CSharp/Compiler%20Breaking%20Changes%20-%20DotNet%2010.md, which includes this compiler change.

@jjonescz Are there associated library changes as well?

**Comment by jkotas** [comment]
> exp.Compile(preferInterpretation: true); // fails at runtime in C# 14

With Native AOT and potentially other targets that do not support compilation, this fails at runtime even with `preferInterpretation: true` omitted.

**Comment by davidroth** [comment]
This is basically a dead end when using classic "EF6" together with latest dotnet/c#.  If you want to continue using the latest dotnet with latest lang versions with EF6, you are stuck at C#13.

**Comment by jjonescz** [comment]
> This is basically a dead end when using classic "EF6" together with latest dotnet/c#.

Not sure what you mean by "dead end". You should be always able to work around this break by casting the array to IEnumerable, or calling Enumerable.Contains as static method, etc.

**Comment by davidroth** [comment]
@jjonescz Yes the `.AsEnumerable` workaround works. However, when you have an app with thousands of queries you have to check all those queries add `AsEnumerable` in the correct places. Sure, I can search for `.Contains(` but I dont see a fully automated solution to just search+replace everything automatically, without manually checking each occurrence. 

But yeah, there is a workaround.

**Comment by jjonescz** [comment]
> I dont see a fully automated solution to just search+replace everything automatically, without manually checking each occurrence.

There might be an analyzer / code fixer written for this.

But we are also considering a language change to avoid the break. (The first-class Span feature that introduced this is still in preview after all.)

**Comment by PaulVrugt** [comment]
> This is basically a dead end when using [classic "EF6" ](https://github.com/dotnet/ef6)together with latest dotnet/c#. If you want to continue using the latest dotnet with latest lang versions with EF6, you are stuck at C#13. (Unless the EfCore Team implements a [similar workaround](https://github.com/dotnet/efcore/commit/6c0106b2693c3d3d87c26d8d63d97ad2236c3f4f) for EF6 too).

So, we are running into this now. Did you ever find a fix for the ef6 issues with this? We have a part of a codebase still using ef6, and things like: dbContext.Entities.Where(x => Ids.Contains(x.Id) fail now with 

```
LINQ to Entities does not recognize the method 'Boolean Contains[Guid](System.ReadOnlySpan`1[System.Guid], System.Guid)' method, and this method cannot be translated into a store expression.
```
this issue is closed because of https://github.com/dotnet/docs/pull/44625 , which implies the breaking change was fixed, but it doesn't seem fixed for ef6 queries. Does this mean we are stuck on langversion 13 for anything that still uses ef6?

**Comment by jjonescz** [comment]
You should be able to explicitly use `Enumerable.Contains(Ids, x.Id)` instead of `Ids.Contains(x.Id)` as a workaround.

**Comment by PaulVrugt** [comment]
@jjonescz yes, but that implies us finding the 1000 places where we do this, and changing the code, which only fails at runtime. This is a very risky operation, so I was hoping for a more robust solution

**Comment by jjonescz** [comment]
I guess it would be best to fix this in EF6, I'm not sure how high the bar for changes there is though.

If there is a single place where the Expression trees go through in your code base (or a place where you can hook in EF6), you could probably rewrite them to replace any uses of `MemoryExtensions.Contains`.

You could also use some analyzer, like [BannedApiAnalyzer](https://github.com/dotnet/roslyn/blob/main/src/RoslynAnalyzers/Microsoft.CodeAnalysis.BannedApiAnalyzers/BannedApiAnalyzers.Help.md) to tell you where you are using `MemoryExtensions.Contains`.

**Comment by roji** [comment]
@PaulVrugt yes - for now, EF6 users are recommended to stay at C# 13. You can upgrade to .NET 10 and use the latest SDK/compiler, but keep your C# language at 13 for now (via the `<LangVersion>` setting in your csproj). Otherwise, either changing queries to use Enumerable.Contains or to use `List<T>` instead dof `T[]` should work.

The bar for any changes in EF6 is very high at this point; however, if enough people are running into this, we'll consider making the changes to make it support Contains on C# 14. We'll gather feedback on this via https://github.com/dotnet/ef6/issues/2322.
